types/src/test/java/org/infinispan/protostream/types/java/TypesMarshallingTest.java

```diff
@@ -17,62 +17,127 @@
+                    case BYTE_ARRAY, INPUT_STREAM, JSON -> Stream.of(new TestConfiguration(t, false, false, null));
+                    default -> Stream.of(
+                            new TestConfiguration(t, true, true, null),
+                            new TestConfiguration(t, true, false, ArrayList::new),
+                            new TestConfiguration(t, true, false, HashSet::new),
+                            new TestConfiguration(t, true, false, LinkedHashSet::new),
+                            new TestConfiguration(t, true, false, LinkedList::new),
+                            new TestConfiguration(t, true, false, TreeSet::new));
+                })
                 .map(t -> new Object[]{t})
                 .toArray(Object[][]::new);
     }
 
     @Test
     public void testUUID() throws IOException {
-        method.marshallAndUnmarshallTest(UUID.randomUUID(), context);
+        testConfiguration.method.marshallAndUnmarshallTest(UUID.randomUUID(), context, false);
     }
 
     @Test
     public void testBitSet() throws IOException {
         var bytes = new byte[ThreadLocalRandom.current().nextInt(64)];
         ThreadLocalRandom.current().nextBytes(bytes);
-        method.marshallAndUnmarshallTest(BitSet.valueOf(bytes), context);
+        testConfiguration.method.marshallAndUnmarshallTest(BitSet.valueOf(bytes), context, false);
     }
 
     @Test
     public void testBigDecimal() throws IOException {
-        method.marshallAndUnmarshallTest(BigDecimal.valueOf(ThreadLocalRandom.current().nextDouble(-256, 256)), context);
+        testConfiguration.method.marshallAndUnmarshallTest(BigDecimal.valueOf(ThreadLocalRandom.current().nextDouble(-256, 256)), context, false);
     }
 
     @Test
     public void testBigInteger() throws IOException {
-        method.marshallAndUnmarshallTest(BigInteger.valueOf(ThreadLocalRandom.current().nextInt()), context);
+        testConfiguration.method.marshallAndUnmarshallTest(BigInteger.valueOf(ThreadLocalRandom.current().nextInt()), context, false);
+    }
+
+    @Test
+    public void testContainerWithString() throws IOException {
+        assumeTrue(testConfiguration.runTest);
+        if (testConfiguration.isArray) {
+            testConfiguration.method.marshallAndUnmarshallTest(stringArray(), context, true);
+        } else {
+            testConfiguration.method.marshallAndUnmarshallTest(stringCollection(testConfiguration.collectionBuilder), context, false);
+        }
+    }
+
+    @Test
+    public void testContainerWithBooks() throws IOException {
+        assumeTrue(testConfiguration.runTest);
+        if (testConfiguration.isArray) {
+            testConfiguration.method.marshallAndUnmarshallTest(bookArray(), context, true);
+        } else {
+            testConfiguration.method.marshallAndUnmarshallTest(bookCollection(testConfiguration.collectionBuilder), context, false);
+        }
+    }
+
+    @Test
+    public void testPrimitiveCollectionCompatibility() throws IOException {
```

Comment:
@ryanemerson I added a test here, so I'm confident it is backward compatible.

Reply:
Great work! So I guess the idea would be in Infinispan that we will need to read old store entries with `wrapCollectionElements(false)` and migrate the data to use `wrapCollectionElements(true)` on initial server startup?

Reply:
You can read the old store entries with `wrapCollectionElements(true)`. Collections in object fields are not affected by this bug. 
The only issue is if the user has a `RemoteCache<String, List<String>>` but that is broken and does not work with the current releases.

All of this just to say setting `wrapCollectionElements(true)` at the server/client in the next Infinispan release should be fine.

Reply:
@pruivo I thought `RemoteCache<String, List<String>>` would work if the list only contained a single entry?

Reply:
Collection of primitive works with any number of elements. The collection of Pojo doesn't and only works with strictly a single element. In any case, the Infinspan server does not manipulate the byte[] sent by the client except for Indexing/Query.

Reply:
and transcoding :slightly_smiling_face:

Anyway, thanks for clarifying. IMO the benefits of fixing this issue far outweigh the consequences of older clients having an incompatibility with a new server version specifically for single entry collections. \\cc @tristantarrant